In [14]:
import utils
from tqdm import tnrange, tqdm_notebook
#from collections import Counter

#viz
#import plotly.graph_objs as go
#import plotly.offline as py

%matplotlib inline
import matplotlib.pyplot as plt
import re
from collections import namedtuple
import numpy as np

# No Matter How You Slice It

In [17]:
inp = utils.get_input(3).splitlines()
inp[-5:]

['#1369 @ 61,754: 14x28',
 '#1370 @ 782,52: 29x21',
 '#1371 @ 622,190: 16x12',
 '#1372 @ 209,614: 21x21',
 '#1373 @ 369,713: 20x29']

each claim spec is: `x, y, h, w`

In [68]:
grid = np.zeros(shape=(9,11))
grid

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [69]:
# #123 @ 3,2: 5x4
grid[2:2+4:, 3:3+5] += 1
grid

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.],
       [0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.],
       [0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.],
       [0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [70]:
claims = []

for claim in inp:
    claims.append([int(i) for i in re.findall("[\d]+", claim)])
    
claims[-5:]

[[1369, 61, 754, 14, 28],
 [1370, 782, 52, 29, 21],
 [1371, 622, 190, 16, 12],
 [1372, 209, 614, 21, 21],
 [1373, 369, 713, 20, 29]]

In [71]:
fabric = np.zeros(shape=(1000,1000))
for claim in claims:
    num, x, y, w, h = claim
    fabric[y:y+h,x:x+w] += 1

In [73]:
len(fabric[fabric>=2])

118858

answer to part 1: `118858`

# Part 2

find the overlapping claim

In [81]:
for claim in claims:
    num, x, y, w, h = claim
    if np.sum(fabric[y:y+h,x:x+w]) == w * h:
        print(num, claim)
        break

1100 [1100, 355, 404, 29, 11]


In [78]:
sum(fabric)

array([ 44.,  44.,  73., 123., 123., 154., 182., 196., 215., 262., 314.,
       359., 370., 385., 384., 354., 344., 375., 352., 400., 387., 387.,
       387., 398., 398., 370., 351., 369., 379., 461., 440., 451., 483.,
       458., 473., 426., 493., 543., 562., 550., 547., 582., 539., 565.,
       628., 590., 634., 634., 634., 610., 570., 530., 516., 543., 586.,
       536., 541., 546., 506., 531., 515., 565., 529., 529., 501., 413.,
       426., 488., 518., 518., 521., 529., 441., 529., 553., 532., 562.,
       577., 562., 593., 566., 541., 548., 567., 598., 668., 648., 660.,
       600., 619., 548., 518., 570., 531., 557., 468., 447., 462., 463.,
       436., 455., 443., 482., 494., 515., 598., 653., 681., 704., 677.,
       666., 636., 567., 594., 596., 563., 550., 454., 472., 428., 421.,
       436., 445., 434., 409., 400., 441., 441., 437., 406., 380., 344.,
       301., 332., 285., 270., 287., 271., 268., 255., 269., 266., 305.,
       322., 322., 369., 371., 376., 428., 448., 44